---
title: "Lecture4"
author: "Badri V"
date: "October 3, 2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Downloads") # Set working directory for all chunks
# setwd("~/Downloads") # Set working directory for a chunk
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

##Install packages as needed
```{r}
install.packages("BiocManager")
BiocManager::install(c("snpStats","SNPRelate"), update=F, ask=F)
```
#Load packages
```{r}
require(snpStats)
require(SNPRelate)
##this is needed for GWAA and GWAS_Manhattan functions to run and plot GWAS
source("GWASfunction.R")
```

# Calculate lambda
```{r cars}

####How to calculate Lambda in GWAS
set.seed(0) # changing the random seed is going to chage the lambda. However, if you keep chanigng it many many times the average lambda is closer to the true lambda. For simplicity we are not doing that here but you're encouranged to play with it.
pvalue <- runif(1000, min=0, max=1)
hist(pvalue)
chisq <- qchisq(1-pvalue,1)
hist(chisq)
abline(v=median(chisq), col="red", lwd=3) #median line

# For z-scores as association, just square them
    # chisq <- data$z^2
        #For chi-squared values, keep as is
        #chisq <- data$chisq
lambda = median(chisq)/qchisq(0.5,1)
message("Observed: ", median(chisq), "\nExpected: ", qchisq(0.5,1),
        "\nLambda: ", lambda)

pvalue1 <- rnorm(1000, 0.4, 0.1)
hist(pvalue1)
chisq1 <- qchisq(1-pvalue1,1)
hist(chisq1)
abline(v=median(chisq1), col="red", lwd=3) #median line
lambda1 = median(chisq1)/qchisq(0.5,1)
message("Observed: ", median(chisq1), "\nExpected: ", qchisq(0.5,1),
        "\nLambda: ", lambda1)
#####
```
```{r}

load("PhenoGenoMapCleaned.RData")
names(genData)
##explore MAP and LIP tables

col.summary(genData$SNP)$MAF[1:50]
head(col.summary(genData$SNP))
head(as.data.frame(genData$SNP))
```

```{r}
###run PCA and vizualize the results
###requires a origin file which lists the country of origin (ethnicity) of participants

if (exists("genofile")) snpgdsClose(genofile)
snpgdsBED2GDS(bed.fn = "combined.bed", bim.fn = "combined.bim",
              fam.fn = "combined.fam", out.gdsfn = "myGDS", cvt.chr = "char")
genofile <- snpgdsOpen("myGDS", readonly = F)

pca <- snpgdsPCA(genofile,  num.thread = 1)
pctab <- data.frame(sample.id = pca$sample.id,
                    PC1 = pca$eigenvect[,1],
                    PC2 = pca$eigenvect[,2],
                    stringsAsFactors = F)
origin <- read.delim("origin.txt", sep = "\t")
origin <- origin[match(pca$sample.id, origin$sample.id),]
pcaCol <- rep(rgb(0,0,0,.3), length(pca$sample.id)) # Set black for chinese
pcaCol[origin$origin == "Indian"] <- rgb(1,0,0,.3) # red for indian
pcaCol[origin$origin == "Malay"] <- rgb(0,.7,0,.3) # green for malay

##run three lines together to have it in one plot.
plot(pctab$PC1, pctab$PC2, xlab = "PC1", ylab = "PC2", col = pcaCol, pch = 16)
abline(h = 0, v = 0, lty = 2, col = "grey")
legend("top", legend = c("Chinese", "Indian", "Malay"), col = 1:3, pch = 16, bty = "n")

```


```{r}
# Choose trait for association analysis, use colnames(genData$LIP) for listing
# NOTE: Ignore the first column of genData$LIP (gender)
target <- "Cholesterol"
phenodata <- data.frame("id" = rownames(genData$LIP),
                        "phenotype" = scale(genData$LIP[,target]), stringsAsFactors = F)
# Conduct GWAS (will take a while)
start <- Sys.time()
GWAA(genodata = genData$SNP, phenodata = phenodata, filename = paste(target, ".txt", sep = ""))
Sys.time() - start # benchmark

############################
# Manhattan plot
GWASout <- read.table(paste(target, ".txt", sep = ""), header = T, colClasses = c("character", rep("numeric",4)))
GWASout$type <- rep("typed", nrow(GWASout))
GWASout$Neg_logP <- -log10(GWASout$p.value)
GWASout <- merge(GWASout, genData$MAP[,c("SNP", "chr", "position")])
GWASout <- GWASout[order(GWASout$Neg_logP, decreasing = T),]
GWAS_Manhattan(GWASout)
#############################
```


Feel free to run a GWAS for any other trait in `colnames(genData$LIP)`.
```{r}

```

